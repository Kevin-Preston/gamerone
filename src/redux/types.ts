import { AuthState } from './auth/types';
import { ProfileState } from './profile/types';
import { SettingsState } from './settings/types';
import { PostState } from './post/types';
import { PostFormState } from './post-form/types';
import { RequestStatusState } from './request-status/types';
import { RouterState } from 'connected-react-router';

export const INIT_STATE = 'root/INIT_STATE';
export const LOAD_STATE = 'root/LOAD_STATE';

export interface InitStateAction {
  type: typeof INIT_STATE;
}

export type RootStateActions = InitStateAction;

export interface RootState {
  Auth: AuthState;
  Profile: ProfileState;
  Settings: SettingsState;
  Post: PostState;
  PostForm: PostFormState;
  RequestStatus: RequestStatusState;
  router: RouterState;
}
