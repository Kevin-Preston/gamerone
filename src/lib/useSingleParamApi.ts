import { useState, useEffect } from 'react';
import { Nullable } from 'interfaces';

type PromiseFunc = (p: any) => Promise<any>;

export default function useSingleParamApi<T>(apiAction: PromiseFunc) {
  const [loading, setLoading] = useState(false);
  const [resolved, setResolved] = useState(null);
  const [error, setError] = useState<Nullable<Error>>(null);
  const [param, setParam] = useState<Nullable<T>>(null);

  useEffect(() => {
    const process = async () => {
      setLoading(true);
      try {
        const resolved = await apiAction(param);
        setResolved(resolved);
        setError(null);
      } catch (e) {
        setError(e);
        setResolved(null);
      }
      setLoading(false);
    };
    if (param) process();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [param]);

  return [{ loading, resolved, error }, setParam] as const;
}
