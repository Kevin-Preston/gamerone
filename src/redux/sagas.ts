import { all } from 'redux-saga/effects';
import authSagas from './auth/saga';
import profileSagas from './profile/saga';
import settingsSagas from './settings/saga';
import postSagas from './post/saga';
import postFormSagas from './post-form/saga';
import miscSagas from './misc/saga';

export default function* rootSaga() {
  yield all([
    authSagas(),
    profileSagas(),
    settingsSagas(),
    postSagas(),
    postFormSagas(),
    miscSagas(),
  ]);
}
