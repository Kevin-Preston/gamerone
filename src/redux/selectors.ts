/**
 * Here comes the combined selectors from different states
 */

import { createSelector } from 'reselect';

import { selectCurrentUser } from './auth/selectors';

import {
  selectCurrentProfileUser,
  selectCurrentProfileNetworks,
  selectCurrentProfileSponsors,
  selectCurrentProfileProducts,
} from './profile/selectors';

import {
  selectSettingsSponsors,
  selectSettingsProducts,
  selectSettingsProfileUser,
  selectProfileSocialNetworks,
} from './settings/selectors';

// determines whether the currently viewing profile is my profile or not
export const selectIsSelfProfile = createSelector(
  [selectCurrentProfileUser, selectCurrentUser],
  (profileUser, authUser) => authUser.id === profileUser?.id,
);

export const selectProfileUser = createSelector(
  [selectCurrentProfileUser, selectSettingsProfileUser],
  (profileUser, settingsUser) =>
    settingsUser.id === profileUser?.id ? settingsUser : profileUser,
);

// resolves social networks to be shown on Profile page
export const selectSocialNetworks = createSelector(
  [
    selectIsSelfProfile,
    selectProfileSocialNetworks,
    selectCurrentProfileNetworks,
  ],
  (isSelfProfile, settingsNetworks, profileNetworks) =>
    isSelfProfile ? settingsNetworks : profileNetworks,
);

// resolves sponsors to be shown on Profile page
export const selectSponsors = createSelector(
  [selectIsSelfProfile, selectSettingsSponsors, selectCurrentProfileSponsors],
  (isSelfProfile, settingsSponsors, profileSponsors) =>
    isSelfProfile ? settingsSponsors : profileSponsors,
);

// resolves products to be shown on Profile page
export const selectProducts = createSelector(
  [selectIsSelfProfile, selectSettingsProducts, selectCurrentProfileProducts],
  (isSelfProfile, settingsProducts, profileProducts) =>
    isSelfProfile ? settingsProducts : profileProducts,
);
