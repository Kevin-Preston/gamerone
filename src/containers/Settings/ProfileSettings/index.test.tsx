import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { store } from 'redux/store';
import ProfileForm from './index';

import { waitForSec } from 'lib/testHelper';

/**
 * Mock input event to fill the form with email and password
 *
 * @return {Object}
 */
const renderProfileForm = () => {
  let utils;
  act(() => {
    utils = render(
      <Provider store={store}>
        <MemoryRouter>
          <ProfileForm />
        </MemoryRouter>
      </Provider>,
    );
  });

  return {
    ...utils,
  };
};

/**
 * Fill in the form and submit the form
 *
 * @param {Object} utils
 * @param {String} firstName
 * @param {String} lastName
 * @param {String} email
 * @param {String} username
 * @param {String} password
 */
const fillAndSubmitForm = async (
  utils,
  {
    firstName,
    lastName,
    email,
    username,
    password = 'password',
    bio,
    birthDate,
    newPassword,
  },
) => {
  const { getByRole, getByLabelText } = utils;

  act(() => {
    fireEvent.change(getByLabelText(/firstname/i), {
      target: { value: firstName },
    });
    fireEvent.change(getByLabelText(/lastname/i), {
      target: { value: lastName },
    });
    fireEvent.change(getByLabelText(/username/i), {
      target: { value: username },
    });
    fireEvent.change(getByLabelText(/email/i), {
      target: { value: email },
    });
    fireEvent.change(getByLabelText(/^bio$/i), {
      target: { value: bio },
    });
    fireEvent.change(getByLabelText(/date-picker/i), {
      target: { value: birthDate },
    });
    fireEvent.change(getByLabelText(/^password/i), {
      target: { value: password },
    });
    fireEvent.change(getByLabelText(/^new-password/i), {
      target: { value: newPassword },
    });
    fireEvent.submit(getByRole('button', { name: 'Save' }));
  });
};

test('checks duplicated email', async () => {
  const { getByText, getByLabelText, queryByLabelText } = renderProfileForm();
  const duplicationMessage = 'This email is already registered';

  // mock out window.fetch for the test
  const jestSpy = jest.spyOn(window, 'fetch').mockImplementationOnce(() => {
    const error = new Error();
    error.code = 409;
    return Promise.reject(error);
  });

  const emailElm = getByLabelText(/email/i);

  // checks error message
  act(() => {
    fireEvent.change(emailElm, { target: { value: 'duplicated@email.com' } });
  });
  await waitForElement(() => getByText(duplicationMessage));
  expect(getByText(duplicationMessage)).toBeInTheDocument();

  // checks whether spinner is invisible
  expect(queryByLabelText(/spinner/i)).toBe(null);

  jestSpy.mockRestore();
});

test('checks duplicated username', async () => {
  const { getByText, getByLabelText, queryByLabelText } = renderProfileForm();
  const duplicationMessage = 'This username is taken';

  // mock out window.fetch for the test
  const jestSpy = jest
    .spyOn(window, 'fetch')
    .mockImplementationOnce(async () => {
      const error = new Error();
      error.code = 409;
      return Promise.reject(error);
    });

  const usernameElm = getByLabelText(/username/i);

  // checks error message
  fireEvent.change(usernameElm, { target: { value: 'duplicated-username' } });
  await waitForElement(() => getByText(duplicationMessage));
  expect(getByText(duplicationMessage)).toBeInTheDocument();

  // checks whether spinner is invisible
  expect(queryByLabelText(/spinner/i)).toBe(null);

  jestSpy.mockRestore();
});

test('updates profile and current user', async () => {
  const utils = renderProfileForm();
  const user = {
    email: 'xyz@domain.com',
    firstName: 'test',
    lastName: 'user',
    username: 'testuser',
    password: 'P@s2wordal',
    bio: '',
    birthDate: '',
    newPaswsword: '',
  };

  // mock out window.fetch for the test
  const spy = jest.spyOn(window, 'fetch').mockImplementation((url) => {
    let success = { status: 'success' };
    if (url.endsWith('profile/settings')) {
      success = user;
    }
    return Promise.resolve({
      status: 200,
      json: () => success,
    });
  });

  await fillAndSubmitForm(utils, user);

  await waitForSec(1000);

  // check token with one from localStorage
  expect(utils.asFragment()).toMatchSnapshot();

  // check user with one from the state
  expect(store.getState().Auth.user).toStrictEqual(user);

  spy.mockRestore();
});
