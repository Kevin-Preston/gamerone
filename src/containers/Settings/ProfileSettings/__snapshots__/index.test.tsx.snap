// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updates profile and current user 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-12 form-box"
    >
      <div
        class="form-key"
      >
        <span />
      </div>
      <div
        class="form-title"
      >
        Profile
      </div>
      <div
        class="form-body"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              aria-label="username"
              class="form-control"
              name="username"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              aria-label="email"
              class="form-control"
              name="email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="firstName"
            >
              First Name
            </label>
            <input
              aria-label="firstName"
              class="form-control"
              name="firstName"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="lastName"
            >
              Last Name
            </label>
            <input
              aria-label="lastName"
              class="form-control"
              name="lastName"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="bio"
            >
              Bio
            </label>
            <div
              class="field"
            >
              <textarea
                aria-label="bio"
                class="form-control"
                name="bio"
              />
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              for="birthDate"
            >
              Birth Date
            </label>
            <div
              class="date-picker"
            >
              <div
                class="date-picker__date-selection"
              >
                <input
                  aria-label="date-picker"
                  class="date-picker__input form-control"
                  name="birthDate"
                  readonly=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              for="websiteUrl"
            >
              Website URL
            </label>
            <input
              aria-label="websiteUrl"
              class="form-control"
              name="websiteUrl"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              aria-label="password"
              class="form-control"
              name="password"
              type="password"
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="newPassword"
            >
              New Password
            </label>
            <input
              aria-label="new-password"
              class="form-control"
              name="newPassword"
              type="password"
            />
          </div>
          <div>
            <button
              class="btn btn-primary"
              disabled=""
              type="submit"
            >
              Save
            </button>
          </div>
        </form>
      </div>
      <div
        class="form-footer"
      >
        <span />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
